dcos-log/api/handlers.go

```diff
@@ -0,0 +1,249 @@
+	"strings"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/dcos/dcos-go/dcos-log/journal/reader"
+)
+
+type rangeHeader struct {
+	Cursor string
+	Skip   int64
+	Num    uint64
+}
+
+func (r *rangeHeader) validateCursor() error {
+	// empty cursor allowed
+	if r.Cursor == "" {
+		return nil
+	}
+
+	cArray := strings.Split(r.Cursor, ";")
+	if len(cArray) != 6 {
+		return fmt.Errorf("Incorrect cursor format. Got %s", r.Cursor)
+	}
+
+	//TODO: add more checks
+	return nil
+}
+
+func parseRangeHeader(h string) (r rangeHeader, err error) {
+	r = rangeHeader{}
+
+	// example entries=cursor[[:num_skip]:num_entries]
+	hArray := strings.Split(h, ":")
+	if len(hArray) != 3 {
+		return r, fmt.Errorf("Unexpected header format. Got `%s`", h)
+	}
+
+	r.Cursor = strings.TrimPrefix(hArray[0], "entries=")
+	if err := r.validateCursor(); err != nil {
+		return r, err
+	}
+
+	r.Skip, err = strconv.ParseInt(hArray[1], 10, 64)
+	if err != nil {
+		return r, err
+	}
+
+	r.Num, err = strconv.ParseUint(hArray[2], 10, 64)
+	if err != nil {
+		return r, err
+	}
+
+	return r, nil
+}
+
+func writeErrorResponse(w http.ResponseWriter, code int, msg string) {
+	logrus.Error(msg)
+	http.Error(w, msg, code)
+}
+
+func advanceCursor(rHeader rangeHeader, j *reader.Reader) error {
+	// find the cursor position to start with considering how many entries to skip
+	// negative value allowed
+	if rHeader.Skip > 0 {
+		if _, err := j.Journal.NextSkip(uint64(rHeader.Skip)); err != nil {
+			return fmt.Errorf("Unable to NextSkip cursor: %s", err)
+		}
+	} else if rHeader.Skip < 0 {
+		// if no cursor passed, move cursor to the very end
+		if rHeader.Cursor == "" {
+			if err := j.Journal.SeekTail(); err != nil {
+				return fmt.Errorf("Unable to SeekTail: %s", err)
+			}
+		}
+		if _, err := j.Journal.PreviousSkip(uint64(-rHeader.Skip)); err != nil {
+			return fmt.Errorf("Unable to PreviousSkip cursor: %s", err)
+		}
+	}
+	return nil
+}
+
+func readJournalHandler(w http.ResponseWriter, req *http.Request, stream bool, contentType reader.ContentType) {
+	var (
+		rHeader rangeHeader
+		err     error
+	)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	reqRange := req.Header.Get("Range")
+	if reqRange != "" {
+		rHeader, err = parseRangeHeader(reqRange)
+		if err != nil {
+			e := fmt.Sprintf("Error parsing header `Range`: %s", err)
+			writeErrorResponse(w, http.StatusBadRequest, e)
+			return
+		}
+
+		if stream && rHeader.Num != 0 {
+			writeErrorResponse(w, http.StatusBadRequest, "Cannot use num with stream")
+			return
+		}
+	}
+
+	// open journal reader
+	var limit *reader.Num
+	if rHeader.Num > 0 {
+		limit = &reader.Num{
+			Value: rHeader.Num,
+		}
+	}
+
+	j, err := reader.NewReader(limit, contentType)
+	if err != nil {
+		e := fmt.Sprintf("Error opening journal reader: %s", err)
+		writeErrorResponse(w, http.StatusInternalServerError, e)
+		return
+	}
+
+	go func() {
+		select {
+		case <-ctx.Done():
+			logrus.Info("Requests fulfilled, closing journal")
```

Comment:
i prefer explicit logrus, since my IDE always tries to import "log".
here is an example from docker code https://github.com/docker/docker/blob/master/plugin/backend.go#L61